jenkinsci/jenkins
# [JENKINS-69955] Make websocket connection idleTimeout configurable

and adjust the default

See [JENKINS-69955](https://issues.jenkins.io/browse/JENKINS-69955).

Since Jetty 10, websocket connections have a [default idle timeout of 30s](https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/jetty-websocket/websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketConstants.java#L32). As opposed to a default [idle timeout of 5 minutes](https://github.com/eclipse/jetty.project/blob/jetty-9.4.48.v20220622/jetty-websocket/websocket-api/src/main/java/org/eclipse/jetty/websocket/api/WebSocketPolicy.java#L81-L86) in Jetty 9. 

This does not play well with the current ping interval mechanism that is used to keep connections active. The ping interval is [30 seconds by default](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/jenkins/websocket/WebSocketSession.java#L58). Just a slight drift between 2 pings is enough have an existing websocket connection that expires..

Hence the proposal here to:

* set the websocket idle timeout of the Jetty websocket container to a reasonable default of 2 x the ping interval (60 seconds)
* allow adjusting the idle timeout via a system property `jenkins.websocket.idleTimeout`

### Testing done

**Test 1**

* Start Jenkins
* Create a websocket agent
* Connect the websocket agent
* Wait for minutes / an hour to make sure that the connection is stable

**Test 2: test of system property**

* Start Jenkins with `-Djenkins.websocket.idleTimeout=15`
* Create a websocket agent
* Connect the websocket agent
* Wait for up to 30s to see that  the connection fails (because clearly the idle timeout is lower than the default ping interval)

### Proposed changelog entries

- Adjust websocket Idle Timeout to 60s seconds by default to avoid "WebSocketTimeoutException: Connection Idle Timeout" issues. Idle timeout is configurable via `jenkins.websocket.idleTimeout=<timeoutInSeconds>`.

### Proposed upgrade guidelines

N/A

### Submitter checklist

  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.

### Desired reviewers

@jglick @olamy 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:



<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7670"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
